internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,384 @@
+			ConnectionRefusedCount: aws.Int64(0),
+			OtherCount:             aws.Int64(0),
+			TimeoutCount:           aws.Int64(0),
+			UnknownHostCount:       aws.Int64(0),
+		},
+	}
+}
+
+// Start starts the loop to send the records.
+func (t *telemetryRecorder) Start() {
+	t.startOnce.Do(func() {
+		go t.start()
+	})
+}
+
+// Stop closes the done channel to stop the loop.
+func (t *telemetryRecorder) Stop() {
+	t.stopOnce.Do(func() {
+		close(t.done)
+	})
+}
+
+// start flushes the record once a minute if telemetry
+// data was recordUpdated.
+func (t *telemetryRecorder) start() {
+	ticker := time.NewTicker(t.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if t.recordUpdated.Load() {
+				t.add(t.cutoff())
+				t.recordUpdated.Store(false)
+				if failedToSend, err := t.send(t.flush()); err != nil {
+					for _, record := range failedToSend {
+						t.add(record)
+					}
+				}
+			}
+		case <-t.done:
+			return
+		}
+	}
+}
+
+// cutoff the current record and swap it out with a new record.
+// Sets the timestamp and returns the old record.
+func (t *telemetryRecorder) cutoff() *xray.TelemetryRecord {
+	snapshot := newTelemetryRecord()
+	snapshot.SetSegmentsSentCount(atomic.SwapInt64(t.record.SegmentsSentCount, 0))
+	snapshot.SetSegmentsReceivedCount(atomic.SwapInt64(t.record.SegmentsReceivedCount, 0))
+	snapshot.SetSegmentsRejectedCount(atomic.SwapInt64(t.record.SegmentsRejectedCount, 0))
+	snapshot.SetSegmentsSpilloverCount(atomic.SwapInt64(t.record.SegmentsSpilloverCount, 0))
+	snapshot.BackendConnectionErrors.SetHTTPCode4XXCount(atomic.SwapInt64(t.record.BackendConnectionErrors.HTTPCode4XXCount, 0))
+	snapshot.BackendConnectionErrors.SetHTTPCode5XXCount(atomic.SwapInt64(t.record.BackendConnectionErrors.HTTPCode5XXCount, 0))
+	snapshot.BackendConnectionErrors.SetTimeoutCount(atomic.SwapInt64(t.record.BackendConnectionErrors.TimeoutCount, 0))
+	snapshot.BackendConnectionErrors.SetConnectionRefusedCount(atomic.SwapInt64(t.record.BackendConnectionErrors.ConnectionRefusedCount, 0))
+	snapshot.BackendConnectionErrors.SetUnknownHostCount(atomic.SwapInt64(t.record.BackendConnectionErrors.UnknownHostCount, 0))
+	snapshot.BackendConnectionErrors.SetOtherCount(atomic.SwapInt64(t.record.BackendConnectionErrors.OtherCount, 0))
+	snapshot.SetTimestamp(time.Now())
+	return snapshot
+}
+
+// add to the queue. If queue is full, drop the head of the queue and try again.
+func (t *telemetryRecorder) add(record *xray.TelemetryRecord) {
+	for {
+		select {
+		case t.queue <- record:
+			return
+		case <-t.done:
+			return
+		default:
+			dropped := <-t.queue
+			if t.logger != nil {
+				t.logger.Debug("queue full, dropping telemetry record", zap.Time("dropped_timestamp", *dropped.Timestamp))
+			}
+		}
+	}
+}
+
+// flush the queue into a slice of records.
+func (t *telemetryRecorder) flush() []*xray.TelemetryRecord {
+	var records []*xray.TelemetryRecord
+	for len(t.queue) > 0 {
+		records = append(records, <-t.queue)
+	}
```

Comment:
Can this carry on forever if records are enqueued at the same rate they're dequeued?

Reply:
It's possible, but I think unlikely with the minute interval. Is there a better way to batch and send the records?